### PR TITLE
first draft of vpn role interface design

### PR DESCRIPTION
**Pull request by richm**

### CODE REVIEW

File: documentation/design_docs/vpn/vpn_role_interface_design.md [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+for host in all_hosts | reject('==', inventory_hostname)
+set remotevars = hostvars[host]
+conn {{ construct name based on local and remote host properties }}
+    left={{ inventory_hostname }}
+    leftid={{ vpn_hostid }}
+    right={{ host }}
+    rightid={{ remotevars.remoteid }}
+```
+
+For the host-to-host cases where we need to specify the relationships explicitly, we
```
Comment (richm) [inline documentation/design_docs/vpn/vpn_role_interface_design.md:238]: @simo5 did I capture the gist of what we were discussing?

Comment (simo5) [reply documentation/design_docs/vpn/vpn_role_interface_design.md:238]: I am not sure, my idea was to have something like this:
- vpn-tunnel [some name] // like "host1-host2", could be autogenerated
  - hosts:
     - host1.my.domain
     - host2.my.domain
  - shared secret: {whatever} // this is optional of course

Then we run the playbook and if both hosts are in the inventory we configure both sides of the tunnel. If only one is we configure only that side and consider the other as "managed by a third party".

If you have more than 2 hosts then you do a full mesh configuration among them, examples:

Full mesh done by configuring individual tunnels:
- vpn-tunnel "all-group1"
  - type: host-to-host
  - hosts:
    - host1.group1.mydomain
    - host2.group1.mydomain
    - host3.group1.mydomain
    - host4.group1.mydomain
  - secret-type: keypair

For each of these hosts in the playbook inventory, the role would (in 3 rounds of connecting back and forth to each host):
- create a key pair and pull the public key into the admin host
- configure 3 tunnels to each of the other hosts using the public keys retrieved from each of the hosts
- test that each connection came up properly

Full mesh done via actual mesh and certs:
- vpn-tunnel "all-group1"
  - type: mesh
  - hosts:
    - host1.group1.mydomain
    - host2.group1.mydomain
    - host3.group1.mydomain
    - host4.group1.mydomain
  - secret-type: certificate
  - ca-certificate: blah
  - name constraints: blah // or whatever other parameter makes sense to match certs

This one would configure each host to accept connections from other hosts with a certificate signed by the provided CA and matching some name restriction (or potentially the whole list of hosts is add to a permit-list and SAN names are checked, whatever works for IPSec.

The main feature is that no matter what underlying technology or topology the declaration of the "vpn-tunnel" is always the same, and only specific properties change to steer the role to apply one or another config.
It also avoids to have multiple configurations for each host you want to add in that need to be kept in sync whenever a connection options needs to be changed (say you want to start excluding a specific cipher or host), instead you have a single configuration even for large groups of host-to-host tunnels.

**Comment by richm** [comment]
@simo5 @mprovenc @letoams @pcahyna @nhosoi updated based on conversations with the team today - ptal

Pavel - under vpn_connections hosts we are using dictionary keys which are not valid Python identifiers - but I think this is ok for this usage because Ansible is doing the same thing.  We will not be doing something like `{{ if conn.bastion1.example.com ... }}` since we will never know in the role code exactly what is the hostname, we will use `{{ if conn[host] }}` which I think is fine.

Another difference with the current proposed spec: https://github.com/linux-system-roles/vpn/pull/2/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R12 `vpn_connections` is a `list`, not a `dict` - this is related to the first point about Python identifiers as dict keys, but again we won't know those names ahead of time so won't use them in contexts where it would be a problem.  But that's not the primary concern.  If the name isn't meaningful, then we are forcing the user to come up with some arbitrary name for the connection.  If using a `list`, the name is optional - for the libreswan case, with this proposal, we'll have to figure out some way to generate a meaningful name to use in the ipsec.conf `conn $name` anyway.  Paul has expressed a convention that each side of the tunnel has a meaningful name e.g. one side uses `conn bastion1-to-bastion2` and the other side uses `conn bastion2-to-bastion1` - I don't think we want the user to have to label each side and associate each host with the label to use when it is the local host vs. when it is the remote host.

**Comment by simo5** [comment]
> Paul has expressed a convention that each side of the tunnel has a meaningful name e.g. one side uses `conn bastion1-to-bastion2` and the other side uses `conn bastion2-to-bastion1` - I don't think we want the user to have to label each side and associate each host with the label to use when it is the local host vs. when it is the remote host.

Simplifying out these silly requirements is what the role should do IMHO.
If a user really want to give a specific name, they should be able to, but otherwise the role should do it for them, hopefully a sensible name is chosen, but I would be personally ok even if the name was a Hash/UUID generated from some of the data.

**Comment by letoams** [comment]
On Oct 26, 2020, at 08:58, Simo Sorce <notifications@github.com> wrote:
> 
> ﻿
> Paul has expressed a convention that each side of the tunnel has a meaningful name e.g. one side uses conn bastion1-to-bastion2 and the other side uses conn bastion2-to-bastion1 - I don't think we want the user to have to label each side and associate each host with the label to use when it is the local host vs. when it is the remote host.
> 
It’s the easiest way to ensure all names are unique without using guids. So they are still still somewhat readable if you later want to check how much traffic is going over a tunnel.



> Simplifying out these silly requirements is what the role should do IMHO.
> 

The role is automation. So naming them is not something the user should do. 

> If a user really want to give a specific name, they should be able to, but otherwise the role should do it for them, hopefully a sensible name is chosen, but I would be personally ok even if the name was a Hash/UUID generated from some of the data.
> 
That’s is unwise if you ever want to look at debugging a host with 50 connections that are all named with guids.

**Comment by simo5** [comment]
> That’s is unwise if you ever want to look at debugging a host with 50 connections that are all named with guids.

Just to be clear, I am perfectly ok with names like bastionA-bastionB, it's just that they should be autogenerated and not a required parameter that the admin must provide explicitly.

**Comment by richm** [comment]
> > That’s is unwise if you ever want to look at debugging a host with 50 connections that are all named with guids.
> 
> Just to be clear, I am perfectly ok with names like bastionA-bastionB, it's just that they should be autogenerated and not a required parameter that the admin must provide explicitly.

It will be tricky to autogenerate such a name if all we have is an FQDN or IP address - e.g. if I'm looking at this on an 80x24 screen/tmux terminal, seeing something like `very.long.fqdn.hostname-very.long.fqdn.another.hostname`
otoh, the user can provide meaningful names if it is worth the pain of adding all of these descriptive tunnel names to the ansible inventory/config.

**Comment by simo5** [comment]
I think the only issues arises if some configuration combination end up trying to create the same name for two distinct tunnels, long names may be inconvenient, but as you say, users can override them if that's a problem to them.

**Comment by richm** [comment]
I would like to merge this doc, if there are no objections.  I'll eventually have to sync the naming with what @mprovenc uses in https://github.com/linux-system-roles/vpn/pull/2

**Comment by richm** [comment]
This is now covered by https://github.com/linux-system-roles/vpn/pull/2 so I'm deleting this PR

**Pull request closed without merging**